src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionTarget.cs

```diff
@@ -137,33 +184,69 @@ public void Render()
                     targetContext.DrawBitmap(RefCountable.CreateUnownedNotClonable(_layer), 1,
                         new Rect(_layerSize),
                         new Rect(Size), BitmapInterpolationMode.LowQuality);
-                
-                
-                if (DrawDirtyRects)
-                {
-                    targetContext.DrawRectangle(new ImmutableSolidColorBrush(
-                            new Color(30, (byte)_random.Next(255), (byte)_random.Next(255),
-                                (byte)_random.Next(255)))
-                        , null, _dirtyRect);
-                }
 
-                if (DrawFps)
+                if (DebugOverlays != RendererDebugOverlays.None)
                 {
-                    var nativeMem = ByteSizeHelper.ToString((ulong)(
-                        (Compositor.BatchMemoryPool.CurrentUsage + Compositor.BatchMemoryPool.CurrentPool)  *
-                                                    Compositor.BatchMemoryPool.BufferSize), false);
-                    var managedMem = ByteSizeHelper.ToString((ulong)(
-                        (Compositor.BatchObjectPool.CurrentUsage + Compositor.BatchObjectPool.CurrentPool) *
-                                                                     Compositor.BatchObjectPool.ArraySize *
-                                                                     IntPtr.Size), false);
-                    _fpsCounter.RenderFps(targetContext, FormattableString.Invariant($"M:{managedMem} / N:{nativeMem} R:{RenderedVisuals:0000}"));
+                    if (captureTiming)
+                    {
+                        var elapsed = StopwatchHelper.GetElapsedTime(startingTimestamp);
```

Comment:
Note that at this point the frame rendering isn't actually completed, the commands are still in-flight on the GPU.

Reply:
Yes, that's more "CPU-side render time" currently. Is there currently any easy way to know when the GPU rendering is completely done? (without waiting for the next buffer swap since there's no point in measuring the wait for a vblank).

Reply:
With Vulkan we could asynchronously wait for a fence. 

With ANGLE / OpenGL we can probably put `glFinish` at the end of the frame if diagnostics output is enabled, but that's less than ideal.

Reply:
Thank you, I had a look at Vulkan fences. The same in OpenGL seems to be sync objects. There's `glFenceSync` available with GL ≥ 3.2 and `eglClientWaitSync` in EGL ≥ 1.5. I'll give them a try, that's uncharted territory to me ;)

Reply:
In general just calling `glFinish` should be enough since it will wait for pending drawing to complete.

But has to be enabled by a separate flag since it will affect FPS for the rest of the app.

Reply:
We can ignore it for now since there are some planned changes for the rendering code.